pyomo/core/tests/unit/test_sos_v2.py

```diff
@@ -0,0 +1,613 @@
+# ******************************************************************************
+# ******************************************************************************
+
+import math
+import pyomo.environ as pyo
+import pyomo.common.unittest as unittest
+from pyomo.opt import check_available_solvers
+
+solver_name = "scip"
+
+solver_available = bool(check_available_solvers(solver_name))
+
+# *****************************************************************************
+# *****************************************************************************
+
+
+class TestSOS(unittest.TestCase):
+    @unittest.skipIf(not solver_available, "The solver is not available.")
+    def test_nonindexedsos(self, show_output: bool = False):
+        "Test non-indexed SOS using a single pyomo Var component."
+
+        opt = pyo.SolverFactory(solver_name)
+
+        # *********************************************************************
+
+        # non-indexed SOS constraints.....
+
+        test_vectors = [
```

Comment:
This is a stylistic issue, but I would greatly prefer if each case were it's own `test_` function and then the variants were derived classes.  That is, something like:
```python
class BaseTests(object):
    abs_tol = 1e-3

    def setupClass(self):
        self.model = pyo.ConcreteModel()
        self.model.x = pyo.Var([1], domain=pyo.NonNegativeReals, bounds=(0, 40))
        #...

    def test_scalar(self):
        def rule_mysos(m):
            return ([m.y[a] for a in m.A], [i + 1 for i, _ in enumerate(m.A)])

        if self.use_rule:
            self.model.mysos = pyo.SOSConstraint(rule=rule_mysos, sos=self.sos)
        else:
            # index is not provided, nor are the weights
            self.model.mysos = pyo.SOSConstraint(var=self.model.y, sos=self.sos)

        opt.solve(model, tee=show_output)
        self.assertEqual(len(self.model.mysos), 1)
        self.assertAlmostEqual(pyo.value(model.OBJ), exp_res, self.abs_tol)
        # Can you also check the variable values??

   def test_scalar_with_index(self):
        #...

class TestSOS1_rule(BaseTests, unittest.TestCase):
    sos = 1
    use_rule = True

class TestSOS1_norule(BaseTests, unittest.TestCase):
    sos = 1
    use_rule = False

class TestSOS2_rule(BaseTests, unittest.TestCase):
    sos = 2
    use_rule = True

class TestSOS2_norule(BaseTests, unittest.TestCase):
    sos = 2
    use_rule = False
```

The reason for this is that it is more clear what happened when there is a test failure (in the current implementation, the test fails, but there is no information in the failure as to which case you were actually running)